cdragontoolbox/hashes.py

```diff
@@ -185,8 +185,13 @@ def check_xdbg_hashes(self, path):
     def check_basenames(self, names):
         """Check a list of basenames for each known subdirectory"""
 
-        dirs = self.directory_list()
-        self.check_iter(f"{d}/{name}" for name in names for d in dirs)
+        dirs = list(self.directory_list())
+        unknown = self.unknown # global -> local for increased performance
+        for name in progress_iterator(sorted(names)):
+            for dir in dirs:
+                h = xxh64(name).intdigest() # inline check() for increased performance
```

Comment:
Optimizations should not affect the readibility and maintainability of this code.
Pure Python code is not meant to, and never will, be incredibly fast. If you aim for speed, you should use another approach/language.

Reply:
I'd argue cause the ``check()`` is only 3 lines and it is the only thing being executed, readability isn't really affected by writing it out like that. And considering that performance is increased by around 30% to before (if I'm not mistaken) I think having it like this is fine.

Reply:
You should fix the checked path (`f"{dir}/{name}"` instead of just `name`) before comparing the performance. ;)

And yes, readibility IS affected because you have 5 lines that use another module and "internal" instance variables instead of just 1 call to a single function with a straightforward name.

Reply:
Checking basenames ``f"{a}.png" for a in wordlist`` results in 22 seconds with a simple ``self.check()``, while having the three lines (and the ``unknown = self.unknown``) in there manually results in 17 seconds. These results are pretty consistent and show that performance increase is indeed at around 30%. Factoring out dots is also a pretty fine thing to do imo, it's even done in ``check_iter()`` for increased performance.

Reply:
So, you have a ~22% performance increase with `f"{a}.png`, which is a simpler use case (shorter string, single format variable). Even if you reach 20% performance increase it's still nothing compared to what you could achieve with a perfomance-oriented language or interpreter.

My point is: if you want performance, aim for it and do it the right way. Instead of manually inlining Python code to gain ~20%, use the proper tools and gain ~200% instead.

Reply:
They are valid methods if your main objective is to improve performance of your Python code. But it's not the case of our code.

Improving performance often comes with a trade-off. And when you inline `check_iter()` and `check()` whenever they are used, the trade-off is not worth it.

Reply:
The main goal is to have automatic guessing for new patches that will find a reasonnably good amount of new hashes in a reasonnable amount of time, while keeping the code clean, readable and maintainable. Here you barely achieve a 20% performance gain (once again, not 30̈% according to your numbers) for an operation that is already fast enough. Notably, the cost of `check_iter()` is nothing compared to the cost of extracting files from WADs to grep them.

Your goal is to bruteforce hashes as fast as possible, it's not the same goal. If you want/need additional guessing methods you can add them to your own script and implement them as you want.

And yes, maintainability and readibility are reduced because checking hashes is not factorized anymore. So if the implementation of `check()`, `check_iter()` or some `HashGuesser' internals change, you'll have to check every function and modify it if needed.

Reply:
You're right about the factorization part; the code will have to be manually adjusted every time the ``check()`` code changes. I find the probability for that to happen to be very low though, and even if we were to make minor adjustments; replacing 6 occurences in the code with these surely isn't gonna be an hourly long process.

The functions I am optimizing in this PR are not functions that are frequently run on new patch releases; I am optimizing functions that are specifically here to provide brute forcing functionality for hashes that can't be found by grepping files. Surely the automatic guessing functionality isn't the only focus, else there would not be other brute forcing functions in the first place.

Also I don't see how you're getting your 20% performance increase value; from recent tests I have a running time of 76 seconds for ``substitute_basenames()`` on the old version, and 59 seconds for my PR'd version. 76/59 ≈ 1.29, which means 29% faster, not 20%.

Reply:
You said time went down from 22s to 17s. Those are the numbers you gave.

Reply:
And 22/17 is 1.294, which again is ~ 30%+.

Reply:
You're right for the numbers, my bad. I wrongly swapped the two values.

Regarding the inlining, it's not about which method is frequently run or not, it's about code clarity, consistency. As I already said, if your goal is to provide functions *specifically here to provide brute forcing*, you are doing it wrong, because performance will be bad anyway compared to what you could **easily** achieve using a suitable language (e.g. C++).

Reply:
> if your goal is to provide functions specifically here to provide brute forcing, you are doing it wrong

We have numerous functions that rely on brute forcing. Of course having this entire thing be written in a language that supports more optimisations like C or C++ is generally a good idea, it's just currently not the case. For the time being I don't any problem regarding brute forcing functions, especially cause they aren't simple "test this string" functions but are rather manipulating already found strings to guess new ones. 

About the code clarity/consistency stuff:
Would you be fine with keeping the self.check_iter(...) in the functions like before but changing the check_iter function to look like this?
```python
    def check_iter(self, paths):
        """Check paths from an iterable"""

        # failsafe for common dumb error
        if isinstance(paths, str):
            raise TypeError("expected iterable of strings, got a string")
        unknown = self.unknown
        for p in paths:
            h = xxh64(p).intdigest()
            if h in unknown:
                self._add_known(h, p)
```

Reply:
Yes. If you want to optimize `check_iter()`, it's fine.